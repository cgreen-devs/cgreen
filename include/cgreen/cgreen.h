#include <stdlib.h>
#include <cgreen/unit.h>
#include <cgreen/suite.h>
#include <cgreen/text_reporter.h>
#include <cgreen/cdash_reporter.h>
#include <cgreen/cute_reporter.h>
#include <cgreen/assertions.h>
#include <cgreen/constraint_syntax_helpers.h>
#include <cgreen/runner.h>
#include <cgreen/boxed_double.h>
