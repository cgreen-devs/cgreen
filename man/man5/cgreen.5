.mso www.tmac
.TH CGREEN 5
.SH NAME
cgreen \- a framework for C and C++ unit tests and mocking
.SH DESCRIPTION
.B Cgreen
is a framework for creating and running compact and easy-to-read
unittests for C and/or C++.

Here are some of its features:

.IP \[bu] 2
Fluent API resulting in very readable tests
.IP \[bu]
Expressive and clear output using the default reporter
.IP \[bu]
Fully functional mocks, both strict and loose
.IP \[bu]
Each test runs in its own process for test suite robustness
.IP \[bu]
Automatic discovery and running of tests using dynamic library inspection
.IP \[bu]
Extensive and expressive constraints for many datatypes
.IP \[bu]
BDD-flavoured test declarations with Before and After declarations
.IP \[bu]
Extensible reporting mechanism
.IP \[bu]
Fully composable test suites
.IP \[bu]
An isolated test can be run in a single process for debugging


.SH "SEE ALSO"
cgreen-runner(1)
.PP
The full documentation for
.B cgreen-runner
and
.B the Cgreen framework
is in the
.B Cgreen
manual available at
.URL https://github.com/cgreen-devs/cgreen GitHub .
