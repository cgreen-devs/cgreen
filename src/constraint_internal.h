#ifndef CONSTRAINT_INTERNAL_H
#define CONSTRAINT_INTERNAL_H

#include <cgreen/constraint.h>

/* constraints internal functions are used from some user level tests so must be compilable in C++ */
#ifdef __cplusplus
namespace cgreen {
    extern "C" {
#endif

extern void destroy_empty_constraint(Constraint *constraint);
extern void destroy_static_constraint(Constraint *constraint);
extern void destroy_string_constraint(Constraint *constraint);
extern void destroy_double_constraint(Constraint *constraint);
extern void destroy_by_value_constraint(Constraint *constraint);
extern void destroy_constraint(Constraint *);
extern void destroy_constraints(va_list constraints);


extern bool no_expected_value_in(const Constraint *constraint);
extern bool values_are_strings_in(const Constraint *constraint);
extern bool is_content_comparing(const Constraint *constraint);
extern bool is_content_setting(const Constraint *constraint);
extern bool is_not_content_setting(const Constraint *constraint);
extern bool is_string_comparing(const Constraint *constraint);
extern bool is_double_comparing(const Constraint *constraint);
extern bool is_comparing(const Constraint *constraint);
extern bool is_not_comparing(const Constraint *constraint);
extern bool is_parameter(const Constraint *);
extern bool constraint_is_not_for_parameter(const Constraint *, const char *);
extern bool constraint_is_for_parameter(const Constraint *, const char *);
extern bool constraint_is_for_parameter_in(const Constraint *, const char *);
extern bool doubles_are_equal(double tried, double expected);
extern bool double_is_lesser(double actual, double expected);
extern bool double_is_greater(double actual, double expected);
extern bool doubles_are_nearly(double tried, double expected);


#ifdef __cplusplus
    }
}
#endif

#endif
