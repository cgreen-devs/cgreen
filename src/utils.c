#include "utils.h"

#include <stdlib.h>
#include <stdarg.h>
#include <stdio.h>
#include <string.h>
#include <cgreen/memory.h>

#include "../gitrevision.h"

char *cgreen_library_version = VERSION;
char *cgreen_library_revision = GITREVISION;

char *string_dup(const char *string) {
    char *dup = (char *)malloc(strlen(string)+1);
    if (dup)
        strcpy(dup, string);
    return dup;
}


static char *panic_message_buffer = NULL;

void panic_set_output_buffer(const char *buffer) {
    panic_message_buffer = (char *)buffer;
}

void panic(const char *filename, int line, const char *fmt, ...) {
    va_list args;
    va_start(args, fmt);
    if (panic_message_buffer == NULL) {
        fprintf(stderr, "CGREEN EXCEPTION: <%s:%d>", filename, line);
        vfprintf(stderr, fmt, args);
    } else {
        sprintf(panic_message_buffer, "CGREEN EXCEPTION: <%s:%d>", filename, line);
        vsprintf(&panic_message_buffer[strlen(panic_message_buffer)], fmt, args);
    }
    va_end(args);
}
