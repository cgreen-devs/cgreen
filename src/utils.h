#ifndef UTILS_HEADER
#define UTILS_HEADER

#include <cgreen/filename.h>
#include <stdbool.h>

#ifdef __cplusplus
namespace cgreen {
    extern "C" {
#endif

        extern bool panic_use_colours;

#define PANIC(...) panic(FILENAME, __LINE__, __VA_ARGS__)
        extern char *string_dup(const char *original);
        extern void panic_set_output_buffer(const char *buffer);
        extern void panic(const char *filename, int line, const char *fmt, ...);

#ifdef __cplusplus
    }
}
#endif

#endif
