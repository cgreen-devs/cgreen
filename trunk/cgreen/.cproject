<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<?fileVersion 5.0.0?>

<cproject storage_type_id="org.eclipse.cdt.core.XmlProjectDescriptionStorage2">
	<storageModule moduleId="org.eclipse.cdt.core.settings">
		<cconfiguration id="cdt.managedbuild.toolchain.gnu.base.1521338569">
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_org.eclipse.cdt.core.settings" id="cdt.managedbuild.toolchain.gnu.base.1521338569" name="Default"/>
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_cdtBuildSystem" version="4.0.0"/>
			<storageModule externalCElementFile="cdt.managedbuild.toolchain.gnu.base.1521338569_org.eclipse.cdt.core.externalSettings"/>
		</cconfiguration>
	</storageModule>
	<storageModule moduleId="cdtBuildSystem" version="4.0.0">
		<project id="cgreen.null.1958735211" name="cgreen"/>
	</storageModule>
	<storageModule moduleId="org.eclipse.cdt.core.LanguageSettingsProviders"/>
	<storageModule moduleId="org.eclipse.cdt.make.core.buildtargets">
		<buildTargets>
			<target name="all" path="" targetID="org.eclipse.cdt.build.MakeTargetBuilder">
				<buildCommand>make</buildCommand>
				<buildArguments/>
				<buildTarget>all</buildTarget>
				<stopOnError>true</stopOnError>
				<useDefaultCommand>false</useDefaultCommand>
				<runAllBuilders>true</runAllBuilders>
			</target>
			<target name="clean" path="" targetID="org.eclipse.cdt.build.MakeTargetBuilder">
				<buildCommand>rm</buildCommand>
				<buildArguments>-rf build</buildArguments>
				<buildTarget/>
				<stopOnError>true</stopOnError>
				<useDefaultCommand>false</useDefaultCommand>
				<runAllBuilders>true</runAllBuilders>
			</target>
		</buildTargets>
	</storageModule>
	<storageModule moduleId="scannerConfiguration">
		<autodiscovery enabled="true" problemReportingEnabled="true" selectedProfileId=""/>
		<scannerConfigBuildInfo instanceId="cdt.managedbuild.toolchain.gnu.base.1521338569;cdt.managedbuild.toolchain.gnu.base.1521338569.1174655132;cdt.managedbuild.tool.gnu.c.compiler.base.349579178;cdt.managedbuild.tool.gnu.c.compiler.input.1942191604">
			<autodiscovery enabled="false" problemReportingEnabled="true" selectedProfileId="org.eclipse.cdt.make.core.GCCStandardMakePerFileProfile"/>
		</scannerConfigBuildInfo>
		<scannerConfigBuildInfo instanceId="cdt.managedbuild.toolchain.gnu.base.1521338569;cdt.managedbuild.toolchain.gnu.base.1521338569.1174655132;cdt.managedbuild.tool.gnu.cpp.compiler.base.1611546606;cdt.managedbuild.tool.gnu.cpp.compiler.input.560078280">
			<autodiscovery enabled="true" problemReportingEnabled="true" selectedProfileId=""/>
		</scannerConfigBuildInfo>
	</storageModule>
</cproject>
