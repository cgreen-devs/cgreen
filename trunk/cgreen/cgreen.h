#include "unit.h"
#include "text_reporter.h"
#include "assertions.h"
#include <stdlib.h>
