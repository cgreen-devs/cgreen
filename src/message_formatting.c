#include <cgreen/constraint.h>
#include <cgreen/message_formatting.h>
#include <cgreen/string_comparison.h>
#include <inttypes.h>
#ifndef __cplusplus
#include <stdbool.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef __ANDROID__
#include "cgreen/internal/android_headers/androidcompat.h"
#endif // #ifdef __ANDROID__

#ifdef _MSC_VER
#include "wincompat.h"
#endif

#include "constraint_internal.h"


// Handling of percent signs
static const char *next_percent_sign(const char *s) {
    return strchr(s, '%');
}

static size_t count_percent_signs(char const *s)
{
    size_t count = 0;
    char const *p = next_percent_sign(s);
    while (p != NULL) {
        count++;
        p = next_percent_sign(p+1);
    }
    return count;
}


static char *copy_while_doubling_percent_signs(char *string, char const *original)
{
    char *destination = string;
    const char *source = original;
    const char *next = next_percent_sign(source);
    for (; next != NULL; next = next_percent_sign(next+1)) {
        size_t len = next - source + 1;
        memcpy(destination, source, len);
        destination += len;
        *destination++ = '%';
        source = next+1;
    }
    strcpy(destination, source);

    return string;
}


static char *double_all_percent_signs_in(const char *original)
{
    size_t percent_count = count_percent_signs(original);
    char *new_string = (char *)malloc(strlen(original) + percent_count + 1);
    if (new_string == NULL) {
        return NULL;
    }
    copy_while_doubling_percent_signs(new_string, original);
    return new_string;
}


static int find_index_of_difference(void *expected, void *actual, size_t size_to_compare) {
    char *expectedp = expected;
    char *actualp = actual;

    while (size_to_compare--) {
        if (*expectedp++ != *actualp++) {
            return (int)((void *)actualp - actual)-1;
        }
    }

    return -1;
}

static bool actual_value_not_necessary_for(Constraint *constraint, const char *actual_string, const char *actual_value_string) {
    (void)constraint; // UNUSED!
    return strings_are_equal(actual_string, actual_value_string) ||
            strings_are_equal(actual_string, "true") ||
            strings_are_equal(actual_string, "false");
}


bool parameters_are_not_valid_for(Constraint *constraint, intptr_t actual) {
    char *message = validation_failure_message_for(constraint, actual);

    bool not_valid = (strlen(message) > 0);

    free(message);

    return not_valid;
}

char *validation_failure_message_for(Constraint *constraint, intptr_t actual) {
    const char *name_has_incorrect_size_message =
        "Wanted to compare contents with [%s],\n"
        "\t\tbut [%ld] was given for the comparison size.";
    const char *null_used_for_compare_message =
        "Wanted to compare contents with [%s],\n"
        "\t\tbut NULL was used for the pointer we wanted to compare to.\n"
        "\t\tIf you want to explicitly check for null, use the is_null constraint instead.";
    const char *null_used_for_actual_message =
        "Wanted to compare contents of [%s] but it had a value of NULL.\n"
        "\t\tIf you want to explicitly check for null, use the is_null constraint instead.";



    size_t message_size = strlen(name_has_incorrect_size_message) +
            strlen(null_used_for_compare_message) +
            strlen(null_used_for_actual_message) +
            512; // just in case

    char *message = (char *)malloc(message_size);
    memset(message, 0, message_size);

//    (void)function; // UNUSED!
//  if (function != NULL && strlen(function) > 0) {
//      snprintf(message, message_size - 1, "\tIn mocked function [%s]:\n", function);
//  } else {
//      snprintf(message, message_size - 1, "\tIn assertion:\n");
//  }

    if (is_content_comparing(constraint)) {
        const char *as_string;
        if (constraint->parameter_name != NULL) {
            as_string = constraint->parameter_name;
        } else {
            as_string = constraint->expected_value_name;
        }

        if ((long signed)constraint->size_of_expected_value <= 0) {
            snprintf(message + strlen(message), message_size - strlen(message) - 1,
                    name_has_incorrect_size_message,
                    as_string,
                    (long signed)constraint->size_of_expected_value);

            return message;
        }

        if ((void *)actual == NULL) {
            snprintf(message + strlen(message), message_size - strlen(message) - 1,
                    null_used_for_compare_message,
                    as_string);

                return message;
        }

        if (constraint->expected_value.value.pointer_value == NULL) {
            snprintf(message + strlen(message), message_size - strlen(message) - 1,
                    null_used_for_actual_message,
                    as_string);

                return message;
        }
    }

    return message;
}


static bool is_equal_to_string_constraint(Constraint *constraint) {
    return strstr(constraint->name, "not ") == NULL || strstr(constraint->name, "equal ") == NULL;
}

/* Formats for printing actual and expected values */
static const char actual_value_string_format[] = "\n\t\tactual value:\t\t\t[\"%s\"]";
static const char expected_value_string_format[] =  "[%s]";
static const char constraint_as_string_format[] = "Expected [%s] to [%s]";
static const char at_offset[] = "\n\t\tat offset:\t\t\t[%d]";
static const char expected_content[] = "\n\t\t\tactual value:\t\t[0x%02x]\n\t\t\texpected value:\t\t[0x%02x]";


static void format_actual_string_value(intptr_t actual_value, char *message, size_t message_size) {
    snprintf(message + strlen(message), message_size - strlen(message) - 1,
             actual_value_string_format,
             (const char *)actual_value);
}


static void format_expected_string_value(Constraint *constraint, char *message, size_t message_size) {
    snprintf(message + strlen(message), message_size - strlen(message) - 1,
             constraint->expected_value_message,
             constraint->expected_value.value.string_value);
}


char *failure_message_for(Constraint *constraint, const char *actual_string, intptr_t actual_value) {
    char actual_int_value_string[32];
    const char *actual_value_as_string;
    char *message;
    size_t message_size = strlen(constraint_as_string_format) +
            strlen(expected_value_string_format) +
            strlen(actual_value_string_format) +
            strlen(at_offset) +
            strlen(constraint->actual_value_message) +
            strlen(constraint->expected_value_message) +
            strlen(constraint->expected_value_name) +
            strlen(constraint->name) +
            strlen(actual_string) +
            512; // just in case

    snprintf(actual_int_value_string, sizeof(actual_int_value_string) - 1, "%" PRIdPTR, actual_value);

    if (values_are_strings_in(constraint)) {
        message_size += strlen(constraint->expected_value.value.string_value);
        if (actual_value != (intptr_t)NULL) {
            message_size += strlen((char *)actual_value);
        }
    }

    message = (char *)malloc(message_size);

    /* if the actual value expression contains '%' we want it to survive the final expansion with
       arguments that happens in assert_true() */
    actual_value_as_string = double_all_percent_signs_in(actual_string);

    /* expand the constraint with the actual value in string format... */
    snprintf(message, message_size - 1,
             constraint_as_string_format,
             actual_value_as_string,
             constraint->name);

    free((void*)actual_value_as_string);

    if (no_expected_value_in(constraint)) {
        return message;
    } else
        strcat(message, " ");

    /* expand the expected value string for all assertions that have one... */
    snprintf(message + strlen(message), message_size - strlen(message) - 1,
             expected_value_string_format,
             constraint->expected_value_name);

    if (actual_value_not_necessary_for(constraint, actual_string, actual_int_value_string)) {
        /* when the actual string and the actual value are the same, don't print both of them */
        /* also, don't print "0" for false and "1" for true */
        /* also, don't print expected/actual for contents constraints since that is useless */
        return message;
    }

    /* for string constraints, print out the strings encountered and not their pointer values */
    if (values_are_strings_in(constraint)) {
        format_actual_string_value(actual_value, message, message_size);
        if (is_equal_to_string_constraint(constraint)) {
            strcat(message, "\n");
            format_expected_string_value(constraint, message, message_size);
        }
        /* The final string may have percent characters, so, since it is
           later used in a (v)printf, we have to double them
        */
        if (next_percent_sign(message) != NULL) {
            char *message_with_doubled_percent_signs = double_all_percent_signs_in(message);
            free(message);
            message = message_with_doubled_percent_signs;
        }
        return message;
    }

    /* show difference for contents as a position */
    if (is_content_comparing(constraint)) {
        int difference_index = find_index_of_difference(constraint->expected_value.value.pointer_value,
                                                        (void *)actual_value,
                                                        constraint->size_of_expected_value);
        if (difference_index != -1) {
            snprintf(message + strlen(message), message_size - strlen(message) - 1,
                     at_offset,
                     difference_index);
            snprintf(message + strlen(message), message_size - strlen(message) - 1,
                     expected_content,
                     ((char *)actual_value)[difference_index],
                     ((char *)constraint->expected_value.value.pointer_value)[difference_index]);
        }
        return message;
    }

    /* add the actual value */
    snprintf(message + strlen(message), message_size - strlen(message) - 1,
             constraint->actual_value_message,
             actual_value);

    /* add the expected value */
    if (strstr(constraint->name, "not ") == NULL) {
        strcat(message, "\n");
        snprintf(message + strlen(message), message_size - strlen(message) - 1,
                 constraint->expected_value_message,
                 constraint->expected_value.value.string_value);
    }

    return message;
}
