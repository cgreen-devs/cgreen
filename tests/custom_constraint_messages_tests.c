#include <cgreen/cgreen.h>
#include <cgreen/message_formatting.h>
#include "constraint_internal.h"
#include "cgreen_value_internal.h"
#include "utils.h"

Describe(CustomConstraint);
BeforeEach(CustomConstraint) {}
AfterEach(CustomConstraint) {}


/* Example 1 from documentation : is_bigger_than_5()

   No argument to the constraint.

 */

bool compare_want_greater_than_5(Constraint *constraint, CgreenValue actual) {
    (void)constraint;
    return actual.value.integer_value > 5;
}

Constraint static_is_bigger_than_5 = {
        /* .type */ CGREEN_VALUE_COMPARER_CONSTRAINT,
        /* .name */ "be bigger than 5",
        /* .destroy */ destroy_static_constraint,
        /* .compare */ compare_want_greater_than_5,
        /* .test */ test_want,
        /* .format_failure_message_for */ failure_message_for,
        /* .actual_value_message */ "",
        /* .expected_value_message */ "",
        /* .expected_value */ {CGREEN_INTEGER, {5}, 0},
        /* .stored_value_name */ "null",
        /* .parameter_name */ NULL,
        /* .size_of_stored_value */ 0,
        /* .side_effect_callback */ NULL,
        /* .side_effect_data */ NULL
};

/* Remember: failing tests to get output */
Ensure(CustomConstraint, custom_constraint_using_static_function) {
    Constraint * is_bigger_than_5 = &static_is_bigger_than_5;
    assert_that(1, is_bigger_than_5);
}


/* Example 2 from documentation : is_smaller_than()

   Standard data type argument to constraint.

 */

bool compare_want_smaller_value(Constraint *constraint, CgreenValue actual) {
    return actual.value.integer_value < constraint->expected_value.value.integer_value ;
}

Constraint *create_smaller_than_constraint(intptr_t expected_value, const char *expected_value_name) {
    Constraint *constraint = create_constraint();

    constraint->expected_value = make_cgreen_integer_value(expected_value);
    constraint->expected_value_name = string_dup(expected_value_name);
    constraint->type = CGREEN_VALUE_COMPARER_CONSTRAINT;

    constraint->compare = &compare_want_smaller_value;
    constraint->execute = &test_want;
    constraint->name = "be smaller than";
    constraint->size_of_expected_value = sizeof(intptr_t);

    return constraint;
}
#define is_smaller_than(value) create_smaller_than_constraint(value, #value)


Ensure(CustomConstraint, custom_constraint_using_a_function_with_arguments_function) {
    assert_that(19, is_smaller_than(10));
}


/* Example 3 from documentation : can_fit_in_box()

   Using custom data types.

*/
typedef struct Struct {
    char id;
    int size;
} Struct;

typedef struct Piece {
    char id;
    int size;
} Piece;

bool compare_piece_and_box_size(Constraint *constraint, CgreenValue actual) {
    return ((Piece *)actual.value.pointer_value)->size
        < ((Struct*)constraint->expected_value.value.pointer_value)->size ;
}

static void test_fit_piece(Constraint *constraint, const char *function_name, CgreenValue actual,
                           const char *test_file, int test_line, TestReporter *reporter) {
    (*reporter->assert_true)(
            reporter,
            test_file,
            test_line,
            (*constraint->compare)(constraint, actual),
            "Piece [%f], does not fit in [%f] in function [%s] parameter [%s]",
            ((Piece *)constraint->expected_value.value.pointer_value)->id,
            ((Struct *)actual.value.pointer_value)->id,
            function_name,
            constraint->parameter_name);
}

Constraint *create_piece_fit_in_box_constraint(intptr_t expected_value, const char *expected_value_name) {
    Constraint *constraint = create_constraint();

    constraint->expected_value = make_cgreen_pointer_value((void*)expected_value);
    constraint->expected_value_name = string_dup(expected_value_name);
    constraint->type = CGREEN_CONTENT_COMPARER_CONSTRAINT;

    constraint->compare = &compare_piece_and_box_size;
    constraint->execute = &test_fit_piece;
    constraint->name = "fit in box";
    constraint->size_of_expected_value = sizeof(intptr_t);

    return constraint;
}
#define can_fit_in_box(box) create_piece_fit_in_box_constraint((intptr_t)box, #box)

Ensure(CustomConstraint, more_complex_custom_constraint_function) {
    Struct box1 = {.id = (char)1, .size = 5};
    Piece piece99 = {.id = (char)99, .size = 6};
    assert_that(&piece99, can_fit_in_box(&box1));
}
