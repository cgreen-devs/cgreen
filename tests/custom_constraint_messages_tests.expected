Running "custom_constraint_messages_tests" (3 tests)...
custom_constraint_messages_tests.c:000: Failure: CustomConstraint -> custom_constraint_using_a_function_with_arguments_function 
	Expected [19] to [be smaller than] [10]

custom_constraint_messages_tests.c:000: Failure: CustomConstraint -> custom_constraint_using_static_function 
	Expected [1] to [be bigger than 5]

custom_constraint_messages_tests.c:000: Failure: CustomConstraint -> more_complex_custom_constraint_function 
	Expected [&piece99] to [fit in box] [&box1]
		at offset:			[0]
			actual value:		[0x63]
			expected value:		[0x01]

  "CustomConstraint": 3 failures in 0ms.
Completed "custom_constraint_messages_tests": 3 failures in 0ms.
